dev-tools/clickana/src/chart.rs

```diff
@@ -0,0 +1,764 @@
+        let Some(start_time) = self.timestamps.iter().min() else {
+            bail!("failed to retrieve start time, timestamp list is empty")
+        };
+        Ok(start_time)
+    }
+
+    fn max(&self) -> Result<&i64> {
+        let Some(end_time) = self.timestamps.iter().max() else {
+            bail!("failed to retrieve end time, timestamp list is empty")
+        };
+        Ok(end_time)
+    }
+
+    fn avg(&self, start_time: &i64, end_time: &i64) -> i64 {
+        (start_time + end_time) / 2
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct XAxisTimestamps {
+    mid_time_label: DateTime<Utc>,
+    pub start_time_label: DateTime<Utc>,
+    pub end_time_label: DateTime<Utc>,
+    start_time_bound: f64,
+    end_time_bound: f64,
+}
+
+impl XAxisTimestamps {
+    fn new(raw_data: &Vec<SystemTimeSeries>) -> Result<Self> {
+        // Retrieve timestamps only to create chart bounds and labels
+        let timestamps = TimeSeriesTimestamps::new(&raw_data);
+        // These timestamps will be used to calculate start and end timestamps in order
+        // to create labels and set bounds for the X axis. As above, some of these conversions
+        // may lose precision, but it's OK as these values are only used to make sure the
+        // datapoints fit within the graph nicely.
+        let start_time = timestamps.min()?;
+        let end_time = timestamps.max()?;
+        let avg_time = timestamps.avg(start_time, end_time);
+
+        let Some(start_time_label) = DateTime::from_timestamp(*start_time, 0)
+        else {
+            bail!(
+                "failed to convert timestamp to UTC date and time;
+        timestamp = {}",
+                start_time
+            )
+        };
+        let Some(end_time_label) = DateTime::from_timestamp(*end_time, 0)
+        else {
+            bail!(
+                "failed to convert timestamp to UTC date and time;
+        timestamp = {}",
+                end_time
+            )
+        };
+        let Some(mid_time_label) = DateTime::from_timestamp(avg_time, 0) else {
+            bail!(
+                "failed to convert timestamp to UTC date and time;
+        timestamp = {}",
+                avg_time
+            )
+        };
+
+        let start_time_bound = *start_time as f64;
+        let end_time_bound = *end_time as f64;
+
+        Ok(Self {
+            mid_time_label,
+            start_time_label,
+            end_time_label,
+            start_time_bound,
+            end_time_bound,
+        })
+    }
+}
+
+#[derive(Debug, PartialEq)]
+struct DataPoints {
+    data: Vec<(f64, f64)>,
+}
+
+impl DataPoints {
+    fn new(timeseries: &Vec<SystemTimeSeries>) -> Self {
+        // These values will be used to render the graph and ratatui
+        // requires them to be f64
+        let data: Vec<(f64, f64)> = timeseries
+            .iter()
+            .map(|ts| {
+                (
+                    ts.time.trim_matches('"').parse::<f64>().unwrap_or_else(
```

Comment:
Why do we need to parse a timestamp into an `f64`? Can't we use an actual time type instead?

Reply:
same as above, the datapoints in the dataset [need to be f64](https://docs.rs/ratatui/latest/ratatui/widgets/struct.Dataset.html#method.data) for ratatui to render them